### ISSUE TITLE
Unity XR Plugin - GLES3 support

### ISSUE THREAD

**Issue by GnoblarAgency** [googlevr/cardboard#40]
Currently, the XR Plugin requires that you use only the GLES2 graphics API. Is it planned to update this to GLES3? 

We previously mixed Cardboard with ARCore but this is no longer possible as ARCore requires only GLES3.

**Comment by jballoffet** [comment]
Thanks for requesting the feature, adding GLES3 support is on our roadmap. Please star or watch this issue to be notified when it becomes available.

**Comment by RajatPatel5** [comment]
I am also facing the same issue with using ARCore and XR plugin in the same project. Please fix this or provide an alternate temporary solution.

**Comment by StefanoCecere** [comment]
top problem. thank you

**Comment by darkverse** [comment]
Yes major problem. OPENGLES3 would be a must have but are there any plans for Vulkan? Is it on the roadmap at all?

**Comment by Roborino** [comment]
This is a huge issue for us.  We can not move forward until this is resolved.  At a min we need XR Plugin to be able to use GLES3 as we are using ARCore with our application.  We are on hold until this can be resolved.

What is the expected ETA?

**Comment by ankityudiz** [comment]
Same Issue faced by me as well in one of our project as we use ARCore plugin and it conflicts with OPENGLES3 graphics API, Can you guys tell us how much time it will still take for an update with GLES3?

**Comment by RajatPatel5** [comment]
I am using ARFoundation as well as Google Cardboard in my project. As this new XR plugin only uses OPENGLE2 which is not supported or either ARkit or ARCore we are unable to use both the plugins in our project. We also tried to use the older plugin but we are not able to resolve the references of UIWebView with it. So please suggest some ways to overcome this issue.

**Comment by hawkwood** [comment]
Instead of this plugin, I am successfully using Mobfish's plugin (https://github.com/mobfishgmbh/Cardboard-VR-Unity-SDK) with ARKit on iOS.  I think the public_dev branch is suppose to be functional with Android, but I haven't tested it yet.

**Comment by maartenslaa** [comment]
Same issue here, I thought OpenGLES 3.0 is backward compatible except for some minor parts: "..framebuffer objects are no longer shared between contexts, cubemaps are always filtered using seamless filtering, and there are minor changes in the way signed fixed-point numbers are converted to floating-point numbers." [source](https://www.informit.com/articles/article.aspx?p=2181697&seqNum=3)

**Comment by Roborino** [comment]
> Thanks for requesting the feature, adding GLES3 support is on our roadmap. Please star or watch this issue to be notified when it becomes available.

Is it possible to get an update?  Our project is currently on hold until this is resolved.

**Comment by Roborino** [comment]
@jballoffet understood that it does not support Vulkan.

Is it possible to get a date when it will support GLES3?

**Comment by chaosemer** [comment]
I expect that adding support for GLES 3.2 is a straightforward shader rewrite. that you can do. We include shaders in native code in distortion_renderer.cc and in cardboard_xr_unity.cc. You could also backport ARCore to GLES2 as there's also nothing major in the implementation tying it to GLES3.2. For ARCore I also expect rewriting the shaders to be GLES2 compatible would be sufficient.

I mention this because we're prioritizing issues such as Issue #11 (Metal support) Issue #35 (Requiring Unity Pro) in front of GLES3 support. I hope this unblocks you. 😄

**Comment by dipaw** [comment]
Any news on OpenGLes 3.0? A lot of developers have been waiting for a long time.

**Comment by chaosemer** [comment]
This was fixed in v1.4.0.

**Comment by dipaw** [comment]
![68e8685b717b71a76c56ca1b13eaf140](https://user-images.githubusercontent.com/22606055/101234031-ffaafb80-36dd-11eb-8d79-4c800fbed5b4.png)
Something goes wrong. :(

**Comment by chaosemer** [comment]
For

**Comment by chaosemer** [comment]
Drat. These type of device-specific issues are super annoying to debug. We test on the Galaxy S8 and S9, but not on the S7 (which your phone appears to be). We noticed a similar issue on those devices in our internal testing but it looks the S7's specific driver doesn't like #version being on the second line, even if the first line is empty.

I'm fairly certain this will be fixed by moving the `#version` lines in opengl_es3_distortion_renderer.cc's [kDistortionVertexShader](https://github.com/googlevr/cardboard/blob/master/sdk/rendering/opengl_es3_distortion_renderer.cc#L37-L38) and [kDistortionFragmentShader](https://github.com/googlevr/cardboard/blob/master/sdk/rendering/opengl_es3_distortion_renderer.cc#L50-L51) and in cardboard_xr_plugin.cc's [kWidgetVertexShaderOpenGlEs3](https://github.com/googlevr/cardboard/blob/master/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc#L198-L199) and [kWidgetFragmentShaderOpenGlEs3](https://github.com/googlevr/cardboard/blob/master/sdk/unity/xr_unity_plugin/cardboard_xr_unity.cc#L211-L212) to be on the very first line, i.e. as `R"glsl(#version 300 es`. Can you try making that change and rebuilding sdk.aar following the instructions in [generate_sdk_aar.md](https://github.com/googlevr/cardboard-xr-plugin/blob/master/Runtime/Android/generate_sdk_aar.md) on your phone (looks like a Samsung Galaxy S7)?

**Comment by dipaw** [comment]
Yes, there really is a mistake on the samsung s7. I checked on the Blackberry Priv, everything starts fine there.
Unfortunately, I don't even have an android studio, and I don't even know how to use it, so I can hardly help you with anything. Sorry.

**Comment by chaosemer** [comment]
What's the build fingerprint on the S8 plus that crashes? (From your log, the S7's build fingerprint is `samsung/heroltexx/herolte:7.0/NRD90M/G93OFXXS1DRA5:user/release-keys`.) We tested on an S8, but it may have been a different OS or chipset. (The S8 has both Qualcomm and Exynos variants and they have completely different GPU behavior.)

**Comment by dipaw** [comment]
![68e8685b717b71a76c56ca1b13eaf140](https://user-images.githubusercontent.com/22606055/101236225-40f8d680-36f1-11eb-9e28-ae98b0ddf3dc.png)

Here is the logcat with S8.

**Comment by chaosemer** [comment]
I created a branch on cardboard-xr-plugin with my hypothetical fix, can you please try that out on your S7 and S8?

The branch is named fix-gles-3.0. You can download it from [here](https://github.com/googlevr/cardboard-xr-plugin/archive/fix-gles-3.0.zip), then unzip it and that package to your Unity project by specifying the unzipped location on disk.

**Comment by dipaw** [comment]
> can you please try that out on your S7 and S8?

Both works now, Thank you 👍

**Comment by chaosemer** [comment]
Glad to hear it! I'm going to open a new bug for the GLES 3.0 doesn't work on some devices and close this one as GLES3 support *is* completed and working otherwise.

**Comment by ArcticWolves** [comment]
will there be support for the Vulkan and  OpenGL 3.0?

**Comment by chaosemer** [comment]
OpenGL ES 3.0 should be working fine in the latest release (1.4.1). If you have specific issues, please file them separately.
